lib/spack/spack/util/package_hash.py

```diff
@@ -74,82 +81,206 @@ def visit_ClassDef(self, node):  # noqa
+
+                    # Boolean literals come through like this
+                    if isinstance(cond, bool):
+                        conditions.append(cond)
+                        continue
+
+                    # otherwise try to make a spec
+                    try:
+                        cond_spec = spack.spec.Spec(cond)
+                    except Exception:
+                        # Spec parsing failed -- we don't know what this is.
+                        conditions.append(None)
+                    else:
+                        # Check statically whether spec satisfies the condition
+                        conditions.append(self.spec.satisfies(cond_spec, strict=True))
+
                 except AttributeError:
                     # In this case the condition for the 'when' decorator is
                     # not a string literal (for example it may be a Python
-                    # variable name). Therefore the function is added
-                    # unconditionally since we don't know whether the
-                    # constraint applies or not.
-                    nodes.append((node, None))
-        else:
-            nodes.append((node, None))
+                    # variable name). We append None because we don't know
+                    # whether the constraint applies or not, and it should be included
+                    # unless some other constraint is False.
+                    conditions.append(None)
+
+        # anything defined without conditions will overwrite prior definitions
+        if not conditions:
+            self.methods[func.name] = []
+
+        # add all discovered conditions on this node to the node list
+        impl_conditions = self.methods.setdefault(func.name, [])
+        impl_conditions.append((func, conditions))
+
+        # don't modify the AST -- return the untouched function node
+        return func
 
 
 class ResolveMultiMethods(ast.NodeTransformer):
-    """Remove methods which do not exist if their @when is not satisfied."""
+    """Remove multi-methods when we know statically that they won't be used.
+
+    Say we have multi-methods like this::
+
+        class SomePackage:
+            def foo(self): print("implementation 1")
+
+            @when("@1.0")
+            def foo(self): print("implementation 2")
+
+            @when("@2.0")
+            @when(sys.platform == "darwin")
+            def foo(self): print("implementation 3")
+
+            @when("@3.0")
+            def foo(self): print("implementation 4")
+
+    The multimethod that will be chosen at runtime depends on the package spec and on
+    whether we're on the darwin platform *at build time* (the darwin condition for
+    implementation 3 is dynamic). We know the package spec statically; we don't know
+    statically what the runtime environment will be. We need to include things that can
+    possible affect package behavior in the package hash, and we want to exclude things
```

Comment:
```suggestion
    possibly affect package behavior in the package hash, and we want to exclude things
```

Reply:
This is a fantastically helpful explanation.

Reply:
I agree, this docstring really helped me understand the issue with multimethods.